libcst/codegen/gen_visitor_functions.py

```diff
@@ -0,0 +1,163 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+import inspect
+from collections import defaultdict
+from typing import Dict, Generator, List, Mapping, Sequence, Set, Type
+
+import libcst as cst
+
+
+DO_NOT_CARE_TYPES: Set[str] = {"CSTNode", "ABC", "object"}
+
+
+def _get_nodes() -> Generator[Type[cst.CSTNode], None, None]:
+    """
+    Grab all CSTNodes that are not a superclass. Basically, anything that a
+    person might use to generate a tree.
+    """
+
+    # pyre-ignore __dir__ exists on all objects.
+    for name in cst.__dir__():
+        if name.startswith("__") and name.endswith("__"):
+            continue
+        if name == "CSTNode":
+            continue
+
+        node = getattr(cst, name)
+        try:
+            if issubclass(node, cst.CSTNode):
+                yield node
+        except TypeError:
+            # This isn't a class, so we don't care about it.
+            pass
+
+
+nodes: Sequence[Type[cst.CSTNode]] = list(_get_nodes())
+classmap: Dict[Type[cst.CSTNode], Set[Type[cst.CSTNode]]] = {}
+for node in nodes:
+    # Map the base classes for this node
+    classmap[node] = set(inspect.getmro(node))
+
+
+def _get_bases(node: Type[cst.CSTNode]) -> Generator[Type[cst.CSTNode], None, None]:
+    for base in classmap[node]:
+        if base.__name__ in DO_NOT_CARE_TYPES:
+            # Ignore bases that are Python implementation details.
+            continue
+        if base not in classmap:
+            # Ignore bases that aren't in our tracking list.
+            continue
+        if base.__name__.startswith("_Base"):
+            # Ignore bases that aren't exported.
+            continue
+        yield base
+
+
+def _get_base(node: Type[cst.CSTNode]) -> Type[cst.CSTNode]:
```

Comment:
Maybe this should be `_get_most_generic_base`? As it is, it's hard to understand the difference between `_get_bases` and `_get_base`.

---

Separately: couldn't you just grab the last exported item in the MRO list before `CSTNode`? Unless I'm misunderstanding MRO (which is quite possible). That seems simpler than recursively traversing base classes.

I came up with this, which seemed to work on every node I tried it on:

```python
def _get_most_generic_base(cls):
    candidates = inspect.getmro(cls)
    
    # We only care about subclasses of CSTNode
    candidates = [base for base in candidates if issubclass(base, cst.CSTNode)]
    # But we don't want CSTNode itself
    candidates = [base for base in candidates if base is not cst.CSTNode]
    # And make sure it's actually an exported base
    candidates = [base for base in candidates if base.__name__ in cst.__all__]
    
    # return the most generic candidate
    return candidates[-1]
```

Reply:
Is that guaranteed by getmro? I'd rather not play code golf here and rewrite everything :\

Reply:
It works for everything I tested it on. I wrote this because I was having a really hard time understanding what the existing code did.

This whole thing is moot though if you make the return type of leave methods `Union[CSTNode, MaybeSentinel, RemovalSentinel]` like I suggest in my later comment.

Reply:
I think if the existing code is fine if the function name is chanced to something like `_get_most_generic_base`. It's not too hard to follow the recursion once you realize what it's trying to do.